File: server/webapp/WEB-INF/rails.new/app/controllers/analytics_controller.rb
##########################################################################
# Copyright 2018 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

class AnalyticsController < ApplicationController
  include AuthenticationHelper

  layout 'single_page_app', only: [:index]

  before_action :check_admin_user_and_403, only: [:index]
  before_action :check_pipeline_exists, only: [:show]
  before_action :check_permissions, only: [:show]

  def index
    @view_title = 'Analytics'
    @pipeline_list = pipeline_configs_service.getGroupsForUser(CaseInsensitiveString.str(current_user.getUsername())).map do |pipelines_config|
      pipelines_config.getPipelines().map(&:getName)
    end.flatten
  end

  def show
    render json: analytics_extension.getAnalytics(params[:plugin_id], params[:type], params[:id], request.query_parameters).toMap().to_h
  rescue => e
    render_plugin_error e
  end

  private

  def check_permissions
    if is_request_for_pipeline_analytics?
      is_analytics_enabled_only_for_admins? ? check_admin_user_and_403 : check_user_can_see_pipeline
    elsif is_request_for_agent_analytics?
      is_analytics_enabled_only_for_admins? ? check_admin_user_and_403 : check_user_and_403
    else
      check_admin_user_and_403
    end
  end

  def render_plugin_error e
    log_java_error(e)
    render :text => "Error generating analytics from plugin - #{params[:plugin_id]}", status: 500
  end

  def log_java_error(e)
    cause = e.getMessage()
    stack_trace = com.google.common.base.Throwables.getStackTraceAsString(e)

    Rails.logger.error "#{cause}:\n\n#{stack_trace}"
  end

  def check_pipeline_exists
    return unless is_request_for_pipeline_analytics?

    pipeline_config = pipeline_config_service.getPipelineConfig(params[:pipeline_name])
    render_message "Cannot generate analytics. Pipeline with name:'#{params[:pipeline_name]}' not found.", 404 if pipeline_config.nil?
  end

  def is_analytics_enabled_only_for_admins?
    system_environment.enableAnalyticsOnlyForAdmins
  end

  def is_request_for_pipeline_analytics?
    params[:type] == 'pipeline'
  end

  def is_request_for_agent_analytics?
    params[:type] == 'agent'
  end
end

File: spark/spark-spa/src/main/java/com/thoughtworks/go/spark/spa/AnalyticsController.java
/*
 * Copyright 2021 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.thoughtworks.go.spark.spa;

import com.google.gson.Gson;
import com.thoughtworks.go.config.PipelineConfigs;
import com.thoughtworks.go.plugin.access.analytics.AnalyticsExtension;
import com.thoughtworks.go.plugin.domain.analytics.AnalyticsData;
import com.thoughtworks.go.server.newsecurity.utils.SessionUtils;
import com.thoughtworks.go.server.service.PipelineConfigService;
import com.thoughtworks.go.spark.Routes;
import com.thoughtworks.go.spark.SparkController;
import com.thoughtworks.go.spark.spring.SPAAuthenticationHelper;
import com.thoughtworks.go.util.SystemEnvironment;
import org.slf4j.LoggerFactory;
import spark.ModelAndView;
import spark.Request;
import spark.Response;
import spark.TemplateEngine;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.thoughtworks.go.spark.Routes.AnalyticsSPA.SHOW_PATH;
import static java.lang.String.format;
import static spark.Spark.*;

public class AnalyticsController implements SparkController {
    private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(AnalyticsController.class);
    private static final Gson GSON = new Gson();

    private final SPAAuthenticationHelper authenticationHelper;
    private final TemplateEngine engine;
    private final SystemEnvironment systemEnvironment;
    private final AnalyticsExtension analyticsExtension;
    private final PipelineConfigService pipelineConfigService;

    public AnalyticsController(SPAAuthenticationHelper authenticationHelper, TemplateEngine engine, SystemEnvironment systemEnvironment, AnalyticsExtension analyticsExtension, PipelineConfigService pipelineConfigService) {
        this.authenticationHelper = authenticationHelper;
        this.engine = engine;
        this.systemEnvironment = systemEnvironment;
        this.analyticsExtension = analyticsExtension;
        this.pipelineConfigService = pipelineConfigService;
    }

    @Override
    public String controllerBasePath() {
        return Routes.AnalyticsSPA.BASE;
    }

    @Override
    public void setupRoutes() {
        path(controllerBasePath(), () -> {
            before("", authenticationHelper::checkAdminUserAnd403);
            get("", this::index, engine);
        });

        path(controllerPath(SHOW_PATH), () -> {
            before("", this::checkPipelineExists, this::checkPermissions);
            before("/*", this::checkPipelineExists, this::checkPermissions);
            get("", this::showAnalytics);
            post("", this::showAnalytics);
        });
    }

    public ModelAndView index(Request request, Response response) {
        HashMap<String, String> locals = new HashMap<String, String>() {{
            List<String> pipelines = new ArrayList<>();
            pipelineConfigService.viewableGroupsFor(SessionUtils.currentUsername()).forEach(
                    (PipelineConfigs config) -> config.getPipelines().forEach(
                            (p) -> pipelines.add(p.name().toString())
                    )
            );
            put("viewTitle", "Analytics");
            put("pipelines", GSON.toJson(pipelines));
        }};
        return new ModelAndView(locals, "analytics/index.ftlh");
    }

    public String showAnalytics(Request request, Response response) {
        try {
            final AnalyticsData analytics = analyticsExtension.getAnalytics(
                    request.params(":plugin_id"),
                    request.params(":type"),
                    request.params(":id"),
                    getQueryParams(request));

            response.type("application/json");
            return GSON.toJson(analytics.toMap());
        } catch (Exception e) {
            LOG.error("Encountered error while fetching analytics", e);
            throw halt(500, format("Error generating analytics from plugin - %s", request.params(":plugin_id")));
        }
    }

    private void checkPipelineExists(Request request, Response response) {
        if (isPipelineRequest(request)) {
            if (null == pipelineConfigService.pipelineConfigNamed(request.queryParams("pipeline_name"))) {
                throw halt(404, format("Cannot generate analytics. Pipeline with name: '%s' not found.", request.queryParams("pipeline_name")));
            }
        }
    }

    private void checkPermissions(Request request, Response response) {
        if (isAnalyticsEnabledOnlyForAdmins()) {
            authenticationHelper.checkAdminUserAnd403(request, response);
            return;
        }

        if (isPipelineRequest(request)) {
            authenticationHelper.checkPipelineViewPermissionsAnd403(request, response);
        } else if (isDashboardRequest(request)) {
            authenticationHelper.checkAdminUserAnd403(request, response);
        } else {
            authenticationHelper.checkUserAnd403(request, response);
        }
    }

    private boolean isPipelineRequest(Request request) {
        return "pipeline".equals(request.params(":type"));
    }

    private boolean isDashboardRequest(Request request) {
        return "dashboard".equalsIgnoreCase(request.params(":type"));
    }

    private boolean isAnalyticsEnabledOnlyForAdmins() {
        return systemEnvironment.enableAnalyticsOnlyForAdmins();
    }

    /**
     * Gets query parameters from request, ignoring duplicates. This
     * is different from Spark's queryMap().toMap()
     *
     * @param request the spark request
     * @return the params as a {@link Map<String, String>}
     */
    private Map<String, String> getQueryParams(Request request) {
        final Map<String, String> queryParams = new HashMap<>();
        request.queryMap().toMap().forEach((k, v) -> {
            queryParams.put(k, v[0]);
        });
        return queryParams;
    }
}
